app/src/generate.py

```diff
@@ -72,11 +73,17 @@ def generate(
             },
         )
 
-    messages.append({"content": query, "role": "user"})
+    # chat_history has the user query as the last item, but we want to insert the context first
+    if chat_history:
+        chat_history.pop()
+        chat_history += messages
+    else:
+        chat_history = messages
```

Comment:
Could this be simplified to:
```suggestion
    if chat_history:
        chat_history.pop()
        messages.extend(chat_history)
```
and revert line 86?

Does the `chat_history` include repeated occurrences of the `system_prompt`?
Adding a unit test would help answer that.

Reply:
chat history doesn't include the `system_prompt` I'll add a test for that.